cpp/src/arrow/util/int_util.h

```diff
@@ -21,6 +21,7 @@
 #include <type_traits>
 
 #include "arrow/status.h"
+#include "arrow/util/int_util_overflow.h"
```

Comment:
I don't know why we should not include `<intsafe.h>` in our public headers but we should not include it here.
Only needed `python/**/*.cc` in #13311 should include it instead.

Reply:
AFAIR, `intsafe.h` includes `Windows.h`.

Reply:
How about creating `cpp/src/arrow/util/slice_util_internal.h` (or something) and move `CheckSliceParams` to the header instead of moving `CheckSliceParams` to `int_util_overflow.h`? I feel that moving `CheckSliceParams` to `int_util_overflow.h` is strange. Because `CheckSliceParams` is a validation function for `Slice`. It's not (almost) related to integer overflow.

Reply:
Sounds fine to me.

Reply:
Great, will do 👍